Lib/asyncore.py

```diff
@@ -650,4 +648,4 @@ def set_file(self, fd):
             self.add_channel()
 
         def fileno(self):
-            return self.socket.fileno()
+            return self._fileno
```

Comment:
I wonder if there is code out there, assuming that file_dispatcher.fileno() raises socket.error(EBADF) when the socket is closed, and this changes the behavior in incompatible way.

So maybe it is better use self.socket.fileno() as you used before, and use another check to find if a dispatcher is closed. For example, we have the unused "closing" flag, that asyncore.dispatcher define, but never use - we can do:

 ```
--- a/Lib/asyncore.py
+++ b/Lib/asyncore.py
@@ -391,6 +391,9 @@ class dispatcher:
         return self._fileno
 
     def close(self):
+        if self.closing:
+            return
+        self.closing = True
         self.connected = False
         self.accepting = False
         self.connecting = False
```

And then use:
   
    if obj.closing:

instead of:

    if obj.fileno() == -1:

It may also be more efficient.